### PR TITLE
[JENKINS-59222] document user on retry

### PR DESCRIPTION
**Pull request by mawinter69**
Adding a UserIdCause to the list of causes. Avoid creating a duplicate CauseAction. Filter out previous users and naginator causes to avoid lengthy cause lists.
The changed way of handling the cause requires to add a BuildTriggerBadgeDeactivator otherwise we will get 2 badges. Store the cause in the action so we can show the cause as tooltip (this includes the build number).

### Testing done
Manual testing as well as adding an automated test

### Submitter checklist

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by ikedam** [comment]
JIRA link: https://issues.jenkins.io/browse/JENKINS-59222

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorBuildTriggerBadgeDeactivator.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.chikli.hudson.plugin.naginator;
+
+import hudson.Extension;
+import hudson.model.Cause;
+import org.jenkinsci.plugins.buildtriggerbadge.provider.BuildTriggerBadgeDeactivator;
+
+@Extension
```
Comment (ikedam) [inline src/main/java/com/chikli/hudson/plugin/naginator/NaginatorBuildTriggerBadgeDeactivator.java:7]: Add `optional=true` as buildtriggerbadge is optional.
(Reference: https://javadoc.jenkins.io/hudson/Extension.html#optional() )

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorAction.java [hunk 2]
```diff
@@ -42,6 +45,16 @@ public NaginatorAction(@CheckForNull Run<?, ?> parentBuild, int retryCount, int
...
 
+    public void setCause(Cause cause) {
+        this.cause = cause;
+    }
+
+    public String getTooltip() {
+        if (cause != null) {
+            return cause.getShortDescription();
+        }
+        return "";
```
Comment (ikedam) [inline src/main/java/com/chikli/hudson/plugin/naginator/NaginatorAction.java:56]: `"Re-scheduled after failure"` for backward compatibility.

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorRetryAction.java [hunk 3]
```diff
@@ -73,12 +77,27 @@ static boolean scheduleBuild(final AbstractBuild<?, ?> build, final int delay, i
...
-        actions.add(build.getAction(CauseAction.class));
+        actions.add(getCauseAction(build, replaceUser, cause));
 
-        return build.getProject().scheduleBuild(delay, new NaginatorCause(build), actions.toArray(new Action[0]));
+        return build.getProject().scheduleBuild2(delay, actions.toArray(new Action[0])) != null;
     }
 
+    private static CauseAction getCauseAction(AbstractBuild<?, ?> build, boolean replaceUser, NaginatorCause cause) {
+        List<Cause> currentCauses = new ArrayList<>(build.getCauses());
+        List<Cause> newCauses = currentCauses.stream().filter(c -> !((c instanceof Cause.UserIdCause) && replaceUser) && !(c instanceof NaginatorCause)).collect(Collectors.toList());
```
Comment (ikedam) [inline src/main/java/com/chikli/hudson/plugin/naginator/NaginatorRetryAction.java:96]: Please add some more tests to improve coverage here.

**Review by ikedam — CHANGES_REQUESTED** [review]
Thanks for the request.
Looks almost good to me. I added some comments to request changes.

Please let me know if you plan more requests. I want to know when I should make a new release.
(If you plan many changes, we should consider adopting this plugin)

**Comment by mawinter69** [comment]
> Thanks for the request. Looks almost good to me. I added some comments to request changes.
> 
> Please let me know if you plan more requests. I want to know when I should make a new release. (If you plan many changes, we should consider adopting this plugin)

This is my last change

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorAction.java (hunk 2 above)
Comment (mawinter69) [reply src/main/java/com/chikli/hudson/plugin/naginator/NaginatorAction.java]: done

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorBuildTriggerBadgeDeactivator.java (hunk 1 above)
Comment (mawinter69) [reply src/main/java/com/chikli/hudson/plugin/naginator/NaginatorBuildTriggerBadgeDeactivator.java]: done

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorRetryAction.java (hunk 3 above)
Comment (mawinter69) [reply src/main/java/com/chikli/hudson/plugin/naginator/NaginatorRetryAction.java:99]: done

**Review by ikedam — APPROVED** [review]
Thanks for updates!
Looks good to me.

**Comment by ikedam** [comment]
I plan to make a new release next weekend: #60

**Comment by ikedam** [comment]
nginator-1.19.1 released: https://github.com/jenkinsci/naginator-plugin/blob/master/CHANGELOG.md#1191

**Pull request merged**